packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,401 @@
+ * presence of the following other rules:
+ *
+ * - oas3-valid-media-example
+ * - oas3-valid-schema-example
+ *
+ * These rules verify that the correct, specific format (date vs date-time) is
+ * used for schemas based on their example value. So, we aren't as specific
+ * with that check in this rule - we recommend either "date" or "date-time".
+ */
+
+module.exports = function (schema, _opts, context) {
+  if (!logger) {
+    ruleId = context.rule.name;
+    logger = LoggerFactory.getInstance().getLogger(ruleId);
+  }
+
+  return checkForDateBasedFormat(
+    schema,
+    context.path,
+    context.documentInventory.resolved
+  );
+};
+
+/**
+ * This function implements a rule that enforces date-based schemas use either
+ * the "date" or "date-time" format, so that they're accurately documented as
+ * date-based logical types. We use a heuristic based on either the name of a
+ * schema (derived from the property name, if the schema is a property schema)
+ * or the example value provided for a given schema or schema property.
+ *
+ * The logic here recursively checks all schemas for the presence of unmarked
+ * date-based schemas. As it traverses the schemas, it compiles a list of
+ * potentially-relevant example values. This way, if an object schema defines
+ * its own example, which includes a value for a nested property that should
+ * be identified by the rule, we can track down the value once we reach the
+ * schema for said property. The logic will also gather any relevant parameter
+ * or media type examples that may be defined outside of the schema path.
+ *
+ * @param {object} s the schema to check
+ * @param {array} p the array of path segments indicating the "location" of the schema within the API definition
+ * @param {object} apidef the resolved API definition
+ * @returns an array containing the violations found or [] if no violations
+ */
+function checkForDateBasedFormat(s, p, apidef) {
+  // Map connecting a list of examples for a schema to its logical path.
+  const examples = {};
+
+  // Check for any examples outside of the schema path - they may be in
+  // request bodies, response bodies, or parameters. Store these separately.
+  const indirectExamples = checkForIndirectExamples(p, apidef);
+
+  return validateNestedSchemas(s, p, (schema, path, logicalPath) => {
+    logger.debug(`${ruleId}: checking schema at location: ${path.join('.')}`);
+    logger.debug(
+      `${ruleId}: logical schema path is : ${logicalPathForLogger(logicalPath)}`
+    );
+
+    // Use a composition-aware utility to gather any examples relevant to this
+    // schema, including those defined on applicator schemas in oneOf, etc.
+    const schemaExamples = getExamplesForSchema(schema);
+    logger.debug(`${ruleId}: ${schemaExamples.length} examples found`);
+
+    if (schemaExamples.length) {
+      // Index the examples with the stringified logical path.
+      examples[logicalPath.join('.')] = schemaExamples;
```

Comment:
Speaking of prepending indirect examples, I think if prepend something like this...

```js
logicalPath.length > 0 ? examples[logicalPath.slice(0, -1).join('.')].map(e => e[logicalPath.at(-1)]).filter(e => e !== undefined) : indirectExamples
```

... then you can probably get rid of most of the logic in `findExample()`, `findValueInExamples()`, and `getObjectValueAtPath()`. I.e., if we eagerly gather available example values from parent examples, we don't have to do so much back-solving to find them later.

Reply:
(Well, you still need to handle `[]` and `*` sentinels, obviously.)

Reply:
This ended up working well! I added this (extending it with comments for readability) and was able to remove a lot of code, including both `findExample()` and `findValueInExamples()`. Note that I needed to keep `getObjectValueAtPath`, as it's used to parse the API definition for indirect examples, but I was able to simplify it to one line now that I don't need to support the sentinel values in it.

Reply:
The one drawback to this is losing specificity in the logs presented to the user but I don't think that's a big deal in this case.